python/mlx/nn/utils.py

```diff
@@ -26,8 +29,37 @@ def inner_fn(params, *args, **kwargs):
 
     value_grad_fn = mx.value_and_grad(inner_fn)
 
+    @wraps(fn)
     def wrapped_value_grad_fn(*args, **kwargs):
         value, grad = value_grad_fn(model.trainable_parameters(), *args, **kwargs)
         return value, grad
 
     return wrapped_value_grad_fn
+
+
+def checkpoint(module: Module, fn: Callable):
```

Comment:
Should we make `fn` `Optional` and use `__call__` if it is not set? It seems like that will be the most common case right?

Reply:
Or maybe better as question: why not hard code `fn` to the module itself and remove it as an argument? Is it a flexibility thing (like I can use `Module.forward`)?

Reply:
Yeah this goes for using `mx.checkpoint` as a decorator above as well. Since the modules are not pure functions, the `nn.checkpoint` is basically a helper for "Checkpoint this function wrt this module's parameters".

Once could presumably want to checkpoint a function even completely outside of the module. Something like the following:

```python
def decode(model, x):
    ... call model(x) a bunch of times ...
```

Btw I agree that `fn` should default to `None` that gets set to `__call__`.